### ISSUE TITLE
Laravel Vapor SQS queue payload "attempts" missing

### ISSUE THREAD

**Issue by booni3** [tenancy/tenancy#134]
This has the same behaviour as tenancy/multi-tenant/issues/855. Root cause may or may not be the same. I will add more detail to this as I investigate.

**Error**
```
Undefined index: attempts
```

**Issue**
- The Laravel Vapor queue driver adds `"attempts" => 0` to the job payload, through the `createPayloadArray` method.
- When T/T is enabled, this method seems to be getting skipped.

**Dirty Fix**
Not really a fix at all - more to just explore the issue further. Once the attempts payload is added into the payload in the queue identification driver, everything else works.

```php
class SaveTenantOnQueuePayload
{
    public function __invoke(string $connection, string $queue = null, array $payload = [])
    {
        $tenant = Tenancy::getTenant();

        $payload = $tenant ? [
            'tenant_key'        => $tenant->getTenantKey(),
            'tenant_identifier' => $tenant->getTenantIdentifier(),
        ] : [];

        return array_merge($payload, ['attempts' => 0]);
    }
}
```

**Further**
I think the main thing at this point is to work out why the Vapor queue `createPayloadArray()` is being skipped. It should be called when a new job is pushed to the queue through any of the queue drivers. For Vapor we should be using sqs.

```php
public function push($job, $data = '', $queue = null)
{
    return $this->pushRaw($this->createPayload($job, $queue ?: $this->default, $data), $queue);
}
```

**Comment by ArlonAntonius** [comment]
Just as a reference.

This is still under investigation. There are a lot of things that need to be checked for us to actually understand what exactly is happening. @booni3 has been an amazing help in all this and even provided a Vapor environment for us :)

Hope to continue the research on this soon.

**Comment by sniper7kills** [comment]
OK; so I did some debugging; I think I have a lead on this...

The `VaporServiceProvider` is extending the default sqs queue.
https://github.com/laravel/vapor-core/blob/dec93b6737578d819e6a29890825f7b90f11dd08/src/VaporServiceProvider.php#L30

But it hasn't been extended when the T/T uses the `QueueManager` class and runs the resolve function.

So when T/T resolves the Queue it receives the `Illuminate\Queue\Connectors\SqsConnector` instead of the intended `Laravel\Vapor\Queue\VaporConnector`

**Comment by booni3** [comment]
I am sure people are going to keep coming across this issue in the future. T/T is now fixed but it can be a hard issue to diagnose. I just fixed an issue where a simple error on a `sprintf` function was causing array to string conversion exception. If this happens in your job before a tenant is identified then you will see the "attempts missing" error.

My advice for Vapor... look into your Cloudwatch Logs (or Sentry/Bugsnag etc. if you have it). Find the specific job that is causing the failure in the payload (it is not always that easy to find!) and go through that job. The "attempts missing" error is simply a result of any exception being thrown prior to the tenant being identified. You will not be able to see the root cause in your logs.

**Comment by booni3** [comment]
I have come across this again with the following scenario:

I am using an elastic search package which extends laravel scout. This builds indexes on the elasticsearch server and must be tenant aware. The main differences from the normal scout setup in the cases of tenant aware models, is we need to provide an search index prefix that identifies the tenant.

In its simplest form, this is just overwriting scouts searchableAs method:

```php
public function searchableAs()
{
    $key = \Tenancy::getTenant()->getTenantKey();
    
    return $key.$this->getTable();
}
```

This is all good and works as expected with Tenancy. The other part of scout is importing records in bulk. Again we need to ensure this is tenant aware. Very simply, we overwrite another scout method that dispatches a tenant aware job instead of a normal one.

```php
public function queueMakeSearchable($models)
{
    dispatch(new MakeTenantSearchable(Tenancy::getTenant(), $models));
}
```

This job, in turn, goes into the package and starts to dispatch more jobs that will perform the action of getting models and sending them to elastic search.

I believe its these inner jobs that have issues. While they are tenant aware, they appear to be bypassing the fix in #137, meaning we once again get an "attempts missing" exception when any exception is thrown, which makes debugging the root of the issue close to impossible.

The above "dirty fix" once again seems to fix this issue, which does show that all jobs are running as expected apart from when there are exceptions.
